contract_comm/election/election.go

```diff
@@ -0,0 +1,208 @@
+	_, err := contract_comm.MakeStaticCall(params.ElectionRegistryId, electionABI, "electValidators", []interface{}{}, &newValSet, params.MaxGasForElectValidators, header, state)
+	if err != nil {
+		return nil, err
+	}
+	return newValSet, nil
+}
+
+type voteTotal struct {
+	Group common.Address
+	Value *big.Int
+}
+
+func getTotalVotesForEligibleValidatorGroups(header *types.Header, state vm.StateDB) ([]voteTotal, error) {
+	var groups []common.Address
+	var values []*big.Int
+	_, err := contract_comm.MakeStaticCall(params.ElectionRegistryId, electionABI, "getTotalVotesForEligibleValidatorGroups", []interface{}{}, &[]interface{}{&groups, &values}, params.MaxGasForGetEligibleValidatorGroupsVoteTotals, header, state)
+	if err != nil {
+		log.Error("DistributeEpochRewards, error calling getTotalVotesForEligibleValidatorGroups", "err", err)
+	}
+
+	voteTotals := make([]voteTotal, len(groups))
+	for i, group := range groups {
+		log.Info("Got group vote total", "group", group, "value", values[i])
+		voteTotals[i].Group = group
+		voteTotals[i].Value = values[i]
+	}
+	return voteTotals, err
+}
+
+func getGroupEpochRewards(header *types.Header, state vm.StateDB, group common.Address, totalEpochRewards *big.Int) (*big.Int, error) {
+	var groupEpochRewards *big.Int
+	_, err := contract_comm.MakeStaticCall(params.ElectionRegistryId, electionABI, "getGroupEpochRewards", []interface{}{group, totalEpochRewards}, &groupEpochRewards, params.MaxGasForGetGroupEpochRewards, header, state)
+	if err != nil {
+		return nil, err
+	}
+	return groupEpochRewards, nil
+}
+
+func DistributeEpochRewards(header *types.Header, state vm.StateDB, groups []common.Address, totalEpochRewards *big.Int) (*big.Int, error) {
+	totalRewards := big.NewInt(0)
+	voteTotals, err := getTotalVotesForEligibleValidatorGroups(header, state)
+	if err != nil {
+		return totalRewards, err
+	}
+
+	rewards := make([]*big.Int, len(groups))
+	for i, group := range groups {
+		reward, err := getGroupEpochRewards(header, state, group, totalEpochRewards)
+		if err != nil {
+			return totalRewards, err
+		}
+		rewards[i] = reward
+	}
+
+	for i, group := range groups {
+		reward := rewards[i]
+		for _, voteTotal := range voteTotals {
+			if voteTotal.Group == group {
+				voteTotal.Value.Add(voteTotal.Value, reward)
+				break
+			}
+		}
+
+		sort.Slice(voteTotals, func(i, j int) bool {
+			return voteTotals[i].Value.Cmp(voteTotals[j].Value) < 0
+		})
+
+		lesser := common.ZeroAddress
+		greater := common.ZeroAddress
+		for i, voteTotal := range voteTotals {
+			if voteTotal.Group == group {
+				if i > 0 {
+					lesser = voteTotals[i-1].Group
+				}
+				if i+1 < len(voteTotals) {
+					greater = voteTotals[i+1].Group
+				}
+				break
+			}
+		}
+		_, err := contract_comm.MakeCall(params.ElectionRegistryId, electionABI, "distributeEpochRewards", []interface{}{group, reward, lesser, greater}, nil, params.MaxGasForDistributeEpochRewards, common.Big0, header, state)
+		if err != nil {
```

Comment:
As we discussed, your plan is to rollback the state if there is an error in the loop.

Reply:
Also, note that you'll need to "rollback" the totalRewards variable as well.

Reply:
> Also, note that you'll need to "rollback" the totalRewards variable as well.

Since we roll this all into `updateValidatorScoresAndDistributeEpochPaymentsAndRewards`, and revert if that fails, we should be doing this, but please double check